[V6] Webapp: Support "broken arrow" button to handle PR hangs

leemgs opened the pull request:
This commit is to provide "broken arrow" button to handle unexpected system
situation such as PR hangs, frequent OOM operations, and too many zombie
processes.

**Changes proposed in this PR:**

* Version 6:
1. Updated variables (e.g., id, ip, password) for code readability

* Version 5:
1. Added a security facility to contol IP address
   - Check if IP address of a system administrator
     is equal to an allowed IP address.

* Version 4:
1. Updated a html design of Broken arrow webapge.

* Version 3:
1. Improved try ... catch statement for readability
2. Fixed incorrect file name : broke_arrow_proceed.php
3. Added comments for maintenance

* Version 2:
1. Added configuration variables in json file
2. Added a routine to handle POST html tag
3. Updated check routine if id and passowr are correct
4. Added doxygtags
5. Specified php files with permission 644

* Version 1:
1. Supported "broken arrow" button

Signed-off-by: Geunsik Lim <geunsik.lim@samsung.com>

leemgs commented on ci/taos/webapp/broken_arrow.php:
```diff
@@ -0,0 +1,15 @@
+
+<form action="broken_arrow_process.php" method="post">
+<li>id: <input type="text" name="id"></li><br>
+<li>password: <input type="text" name="password"></li><br>
```
The current PR is a skeleton. This commit will be improved  at the version 2.

leemgs commented:
**Changes proposed in this PR:**

* Version 2:
1. Added configuration variables in json file
2. Added a routine to handle POST html tag
3. Updated check routine if id and passowr are correct
4. Added doxygtags
5. Specified php files with permission 644

* Version 1:
1. Supported "broken arrow" button

ohsewon commented on ci/taos/webapp/broken_arrow.php:
```diff
@@ -0,0 +1,25 @@
+</ul>
+<br><br>
+
+<form action="broken_arrow_process.php" method="post">
```
typo 'broken_arrow_process.php' -> 'broken_arrow_proceed.php'

leemgs replied:
Nice catch.

ohsewon commented on ci/taos/webapp/broken_arrow_proceed.php:
```diff
@@ -0,0 +1,57 @@
+// Run the "BROKEN ARROW" command.
+if ( $id == $ba_id && $pass == $ba_pass) {
+    //$cmd = "cat $password | su -c 'shutdown -r now'";
+    $cmd = "sudo shutdown -r now";
```
what is different with 'sudo reboot -f'?

leemgs replied:
difference is a security. As you know, we have to do not use the password of system account directly.

ohsewon commented on ci/taos/webapp/broken_arrow_proceed.php:
```diff
@@ -0,0 +1,57 @@
+        echo shell_exec($cmd);
+        exec($cmd, $output);
+        system($cmd, $output);
+        }Catch(Exception e){
```
Catch -> catch, IMO, it is better that adding space between '}' and 'c'

leemgs replied:
Fixed.

leemgs commented:
* Web UI: 

![image](https://user-images.githubusercontent.com/82404/48468788-2c0e2980-e830-11e8-8178-8f1a1e5093b6.png)

leemgs commented:
**Changes proposed in this PR:**

* Version 3:
1. Improved try ... catch statement for readability
2. Fixed incorrect file name : broke_arrow_proceed.php
3. Added comments for maintenance

* Version 2:
1. Added configuration variables in json file
2. Added a routine to handle POST html tag
3. Updated check routine if id and passowr are correct
4. Added doxygtags
5. Specified php files with permission 644

* Version 1:
1. Supported "broken arrow" button

leemgs commented on ci/taos/webapp/broken_arrow_proceed.php:
```diff
@@ -0,0 +1,58 @@
+else {
+    echo ("Oooops. ID and password is incorrect. Please check it.");
+}
+?>
```
Fixed.

leemgs commented:
**Changes proposed in this PR:**

* Version 5:
1. Added a security facility to contol IP address
   - Check if IP address of a system administrator
     is equal to an allowed IP address.

* Version 4:
1. Updated a html design of Broken arrow webapge.

* Version 3:
1. Improved try ... catch statement for readability
2. Fixed incorrect file name : broke_arrow_proceed.php
3. Added comments for maintenance

* Version 2:
1. Added configuration variables in json file
2. Added a routine to handle POST html tag
3. Updated check routine if id and passowr are correct
4. Added doxygtags
5. Specified php files with permission 644

* Version 1:
1. Supported "broken arrow" button

ohsewon commented on ci/taos/webapp/broken_arrow_proceed.php:
```diff
@@ -0,0 +1,91 @@
+    $cmd = "sudo shutdown -r now";
+    $output = array();
+    try{
+         echo shell_exec($cmd);
```
need to keep indentation.

leemgs replied:
Fixed

myungjoo commented on ci/taos/webapp/broken_arrow_proceed.php:
```diff
@@ -0,0 +1,91 @@
+// Security: Check if IP address of a system administrator is equal to allowed IP address.
+$ipaddress = get_client_ip();
+if ( $ba_ipaddress != "" && $ba_ipaddress != $ipaddress )
+   die ("Unable to do the broken arrow. Specify IP address in the configuration file.")
```
```$ba_ipaddress``` ? ```$ba_ip```?

Maybe you need to turn on strict variable declaration enforcing.

leemgs replied:
Fixed.

leemgs commented:
**Changes proposed in this PR:**

* Version 6:
1. Updated variables (e.g., id, ip, password) for code readability

* Version 5:
1. Added a security facility to contol IP address
   - Check if IP address of a system administrator
     is equal to an allowed IP address.

* Version 4:
1. Updated a html design of Broken arrow webapge.

* Version 3:
1. Improved try ... catch statement for readability
2. Fixed incorrect file name : broke_arrow_proceed.php
3. Added comments for maintenance

* Version 2:
1. Added configuration variables in json file
2. Added a routine to handle POST html tag
3. Updated check routine if id and passowr are correct
4. Added doxygtags
5. Specified php files with permission 644

* Version 1:
1. Supported "broken arrow" button

leemgs commented:
The local test is successfully done. This PR (v6) is ready to be merged.
